### PR TITLE
Add IP-Adapter to StableDiffusionXLControlNetImg2ImgPipeline

### PR DESCRIPTION
**Pull request by antoine-scenario**
Add IP-Adapter to StableDiffusionXLControlNetImg2ImgPipeline as it was missing so far.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by charchit7** [comment]
@antoine-scenario run `make quality`

**Comment by antoine-scenario** [comment]
> @antoine-scenario run `make quality`

done 👍

**Comment by patrickvonplaten** [comment]
@yiyixuxu could you take a look here?

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py [hunk 1]
```diff
@@ -197,10 +208,19 @@ class StableDiffusionXLControlNetImg2ImgPipeline(
             Whether to use the [invisible_watermark library](https://github.com/ShieldMnt/invisible-watermark/) to
             watermark output images. If not defined, it will default to True if the package is installed, otherwise no
             watermarker will be used.
+        feature_extractor ([`~transformers.CLIPImageProcessor`]):
+            A `CLIPImageProcessor` to extract features from generated images; used as inputs to the `safety_checker`.
     """
 
     model_cpu_offload_seq = "text_encoder->text_encoder_2->unet->vae"
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:215]:
```suggestion
    model_cpu_offload_seq = "text_encoder->text_encoder_2->image_encoder->unet->vae"
```

**Review by yiyixuxu — APPROVED** [review]
looking great! thank you :)

**Comment by yiyixuxu** [comment]
can we fix the test?

**Comment by antoine-scenario** [comment]
> can we fix the test? you need to add the optional image_encoder component (`None`) here
> 
> https://github.com/huggingface/diffusers/blob/2db73f4a509864d38c668979c58ce1cee57a84db/tests/pipelines/controlnet/test_controlnet_sdxl_img2img.py#L130

@yiyixuxu I did the fix you suggested but the test still fails 😕

**Comment by yiyixuxu** [comment]
ok can we find out what's the difference? I think you should add some print lines here https://github.com/huggingface/diffusers/blob/79c380bc80051d8f82a38c3a7df6f8f4efd1633d/tests/pipelines/test_pipelines_common.py#L570

i.e. print out `pipe.components.keys())` and `init_components.keys()`

**Comment by antoine-scenario** [comment]
> ok can we find out what's the difference? I think you can add some print lines here to find out (i.e. print out `pipe.components.keys())` and `init_components.keys()`)
> 
> https://github.com/huggingface/diffusers/blob/79c380bc80051d8f82a38c3a7df6f8f4efd1633d/tests/pipelines/test_pipelines_common.py#L570

do you have a command to run the test locally ? I followed the instructions from [the doc](https://huggingface.co/docs/diffusers/v0.3.0/en/conceptual/contribution) and tried `pytest tests/pipelines/test_pipelines_common.py` but everything is being skipped.

**Comment by yiyixuxu** [comment]
@antoine-scenario 
oh you should run this test file `pytest tests/pipelines/controlnet/test_controlnet_sdxl_img2img.py`

**Comment by yiyixuxu** [comment]
cc @sayakpaul for a final review

**Review by sayakpaul — APPROVED** [review]
Very nice. 

**_Brainstorming, not related to the PR_**: I wonder if we should work on a tutorial showing the different use cases of IP Adapters in light of various pipelines for which it's supported. 

IP Adapters are getting increasingly popular, so I definitely think it's a good idea. @stevhliu WDYT?

**Comment by charchit7** [comment]
> Very nice.
> 
> **_Brainstorming, not related to the PR_**: I wonder if we should work on a tutorial showing the different use cases of IP Adapters in light of various pipelines for which it's supported.
> 
> IP Adapters are getting increasingly popular, so I definitely think it's a good idea. @stevhliu WDYT?

Yes, IP Adapters quality is really good.

**Comment by stevhliu** [comment]
> IP Adapters are getting increasingly popular, so I definitely think it's a good idea. @stevhliu WDYT?

We currently have a section for IP-Adapters [here](https://huggingface.co/docs/diffusers/main/en/using-diffusers/loading_adapters#other-pipelines) where we show off how to use it with various pipelines and for different tasks as well, so I don't necessarily think we need a new doc for it. But I think we can spin this section out into a standalone doc to increase visibility and keep adding to it. 🙂

**Pull request merged**